File: BOJ/[2800] 괄호 제거/Johoseong/2800.py
import sys
from itertools import combinations
input = sys.stdin.readline        

arr = input()
index, stack = [], []
combination = []
result = set() # 중복 존재하는 경우 제거하기 위해서

#괄호끼리 짝지음
for i, a in enumerate(arr):
    if a == '(':
        stack.append(i)
    if a == ')':
        index.append([stack.pop(), i])

#괄호쌍들 조합
for i in range(1, len(index) + 1):
    combination.extend(list(combinations(index, i)))

for case in combination:
    tmp = list(arr)
    for i in case:
        for j in i:
            tmp[j] = ''
    result.add(''.join(tmp).rstrip())
for i in sorted(result):
    print(i)

File: BOJ/[2800] 괄호 제거/woong-jae/2800.js
const readline = require('readline');
const input = [];

readline.createInterface({
    input: process.stdin,
    output: process.stdout
}).on('line', (line) => {
    input.push(line);
}).on('close', () => {
    console.log(solution(...input));
    process.exit();
});

function solution(s) {
    const braces = [];

    const stack = [];
    for (let i = 0; i < s.length; i++) {
        const char = s[i];
        if (char === "(") {
            stack.push(i);
        }
        else if (char === ")") {
            const front = stack.pop();
            braces.push([front, i]);
        }
    }

    const combinations = [];
    const dfs = (i, picked) => {
        if (i >= braces.length) {
            combinations.push([...picked]);
            return;
        };

        picked.push(i);
        dfs(i + 1, picked);
        picked.pop();

        dfs(i + 1, picked);
    }
    dfs(0, []);

    combinations.pop();

    const removeBrace = (s, braceIndexes) => {
        s = s.split("");
        braceIndexes.forEach(index => {
            const [front, back] = braces[index];
            s[front] = null;
            s[back] = null;
        });
        return s.filter(char => char).join("");
    }

    return Array.from(
        new Set(combinations.map(braceIndexes => removeBrace(s, braceIndexes)))
    ).sort().join("\n");
}